### ISSUE TITLE
Does not work properly with the BelongsToMany relation

### ISSUE THREAD

**Issue by hivokas** [spatie/laravel-query-builder#441]
Recently I've faced a tricky issue with the BelongsToMany relation.

Let's assume we have two models: `Member`, `Team` (many-to-many).

Here is the definition of the `members()` relation for the `Team` model.

```php
public function members()
{
    return $this->belongsToMany(Team::class)->withPivot('joined_at');
}
```

`QueryBuilder::for()` expects `Illuminate\Database\Eloquent\Builder` to be passed.

I want to query the first member of a certain team (including the `joined_at` value), so:

```php
$membersQuery = $team->members();
```

If we call `->first()` - the `pivot` property (instance of `Illuminate\Database\Eloquent\Relations\Pivot`) with the `joined_at` property will be available.

```php
$firstMember = $membersQuery->first();

$joinedAt = $firstMember->pivot->joined_at;
```

Cool.

But if I want to take advantage of this package here - I need to pass the `Illuminate\Database\Eloquent\Builder` instance as I said previously.

So what's the problem?

The problem is that **BelongsToMany** extends **Relation**, while **Relation** does not extend eloquent builder.

So let's just grab this instance from the `$membersQuery`. Let's try.

```php
$membersQueryEloquentBuilder = $membersQuery->getQuery();
```

Now we can pass it to the `QueryBuilder::for()` apply some sort and filters and get the first member.

```php
$firstFilteredAndSortedMember = QueryBuilder::for($membersQueryEloquentBuilder)
    // apply some sorts
    // apply some filters
    ->first();
```

Let's get the pivot instance.

```php
$pivot = $firstFilteredAndSortedMember->pivot; // null
```

Whoops, it is null.

Why? Because logic for retrieving the pivot values is placed inside the `BelongsToMany` relation. But we passed the eloquent builder instance. So we lost this part of logic (which had to retrieve pivot values for us).

After some brainstorming, I came up with a workaround which I placed inside the project. But I'd like to have an opportunity to work with any kind of relation queries out-of-the-box. 

**I could work on the pull request.** But we might need to get rid of the eloquent builder inheritance for `Spatie\QueryBuilder\QueryBuilder` (probably that's a breaking change).

Would you like me to work on that?

**Comment by AlexVanderbist** [comment]
Excellent suggestion, let's continue this discussion in the PR :)

**Comment by malinushj** [comment]
@AlexVanderbist Am I good to just use the v3 branch for now until v3 is released? Any possible timeline on that? (on 2.8.0 at the moment).
Many thanks @hivokas for the PR!! 🎉
